extensions/positron-supervisor/src/Comm.ts

```diff
@@ -1,23 +1,224 @@
+import { JupyterCommMsg } from './jupyter/JupyterCommMsg';
+import { CommMsgRequest } from './jupyter/CommMsgRequest';
+import { Receiver } from './Channel';
+import { CommBackendMessage } from './positron-supervisor';
+import { CommCloseCommand } from './jupyter/CommCloseCommand';
+
+export class CommError extends Error {
+	constructor(
+		message: string,
+		public readonly method?: string
+	) {
+		super(message);
+		this.name = 'CommError';
+	}
+}
+
+export class CommClosedError extends CommError {
+	constructor(commId: string, method?: string) {
+		super(`Communication channel ${commId} is closed`, method);
+		this.name = 'CommClosedError';
+	}
+}
+
+export class CommRpcError extends CommError {
+	constructor(
+		message: string,
+		public readonly code: number = -32000,
+		method?: string
+	) {
+		super(message, method);
+		this.name = 'CommRpcError';
+	}
+}
+
+export class CommImpl implements vscode.Disposable {
+	private readonly disposables: vscode.Disposable[] = [];
+	private closed = false;
+
 	constructor(
 		public readonly id: string,
-		public readonly target: string) {
+		private readonly session: KallichoreSession,
+		public readonly receiver: Receiver<CommBackendMessage>,
+	) { }
+
+	notify(method: string, params?: Record<string, unknown>): void {
+		if (this.closed) {
+			throw new CommClosedError(this.id, method);
+		}
+
+		const msg: CommRpcMessage = {
+			jsonrpc: '2.0',
+			method,
+			params,
+		};
+
+		// We don't expect a response here, so `id` can be created and forgotten
+		const id = createUniqueId();
+		this.session.sendClientMessage(this.id, id, msg);
+	}
+
+	async request(method: string, params?: Record<string, unknown>): Promise<any> {
+		if (this.closed) {
+			throw new CommClosedError(this.id, method);
+		}
+
+		const id = createUniqueId();
+
+		const msg: CommRpcMessage = {
+			jsonrpc: '2.0',
+			id,
+			method,
+			params,
+		};
+
+		const commMsg: JupyterCommMsg = {
+			comm_id: this.id,
+			data: msg
+		};
+
+		const request = new CommMsgRequest(id, commMsg);
+		const reply = await this.session.sendRequest(request);
+
+		if (reply.data.error !== undefined) {
+			const payload = reply.data as CommRpcMessageError;
+			throw new CommRpcError(
+				payload.error.message,
+				payload.error.code,
+				method
+			);
+		}
+
+		if (reply.data.result === undefined) {
+			throw new Error(`Internal error in ${this.id}: undefined result for request ${msg}`);
+		}
+
+		return reply.data.result;
 	}
+
+	close() {
+		this.closed = true;
+	}
+
+	closeAndNotify() {
+		if (this.closed) {
+			return;
+		}
+
+		this.close();
+		const commClose = new CommCloseCommand(this.id);
+		this.session.sendCommand(commClose);
+	}
+
+	// Make sure not to call `dispose()` from Kallichore, only the owner of the
+	// comm should dispose of it. Kallichore calls the `close()` method instead.
+	async dispose(): Promise<void> {
+		this.close();
+		for (const disposable of this.disposables) {
+			disposable.dispose();
```

Comment:
Nit: clear array of disposables after they are disposed, to clear references and make this idempotent